ADC_Sampler/Core/Src/main.c

```diff
@@ -0,0 +1,336 @@
+  *
+  * This software is licensed under terms that can be found in the LICENSE file
+  * in the root directory of this software component.
+  * If no LICENSE file comes with this software, it is provided AS-IS.
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+ADC_HandleTypeDef hadc1;
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+static void MX_GPIO_Init(void);
+static void MX_ADC1_Init(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+uint16_t average_filter(uint16_t val, uint16_t* buffer, uint8_t size) {
+  static uint8_t i = 0;
+  static uint32_t sum = 0;
+
+  sum += val;
+  sum -= buffer[i];
+  buffer[i] = val;
+  i = (i + 1) % size;
+
+  return sum / size;
+}
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+
+  /* USER CODE BEGIN 1 */
+
+  uint16_t raw;
+  uint16_t buffer[5];
+  uint16_t filter;
+
+  for (int i = 0; i < 5; i++) {
+    buffer[i] = 0;
+  }
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_ADC1_Init();
+  /* USER CODE BEGIN 2 */
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+    /* USER CODE END WHILE */
+
+    HAL_ADC_Start(&hadc1);
+    HAL_ADC_PollForConversion(&hadc1, HAL_MAX_DELAY);
+    raw = HAL_ADC_GetValue(&hadc1);
+
+    filter = average_filter(raw, buffer, 5);
+
+    if (filter > 12) {
+      HAL_GPIO_WritePin(GPIOA, GPIO_PIN_6, GPIO_PIN_SET);
+      HAL_GPIO_WritePin(GPIOA, GPIO_PIN_6, GPIO_PIN_RESET);
+    }

```

Comment:
Is this code meant to just turn on and off A6 ( I assume an LED hooked up to it) if the ADC samples 12V? Code runs pretty fast. What's slower, your eyes or the code?

Reply:
Avoid magic numbers (12), if you have random numbers you should have them as variables or constants. It'll save you time banging your head against the wall only to realize you forgot to change a 12 to a 3 somewhere in the code.

Reply:
Don't apologize for mistakes! tbh i'd rather y'all pass in an array for the moving average and then compute it based on that.

I also found an ok version on github: https://gist.github.com/bmccormack/d12f4bf0c96423d03f82
https://github.com/lhr-solar/BPS/pull/598/files here's how BPS handles smth specific with temperature.

Also, I noticed there's a lot of switching between types of integers, uint8_t, uint32_t, etc. Def try to keep it consistent when you can (you can do that in most places since the numbers you're dealing with are small).